File: Top_100_Liked_Questions/560._Subarray_Sum_Equals_K/solution.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        # using hash and seeking for diff
        d = {0:1}
        curr_sum = 0
        sum_minus_k = 0
        count = 0
        for num in nums:
            curr_sum += num
            sum_minus_k = curr_sum - k
            if sum_minus_k in d:
                count += d[sum_minus_k]
            d[curr_sum] = d.get(curr_sum, 0) + 1
        return count
    
    
    def subarraySum2(self, nums: List[int], k: int) -> int:
        ans = 0
        for i in range(len(nums)-1, -1, -1):
            curr_sum = 0
            for j in range(i, len(nums)):
                curr_sum += nums[j]
                if curr_sum == k:
                    ans += 1
        
        return ans

File: Top_100_Liked_Questions/560._Subarray_Sum_Equals_K/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    let hash = {0: 1};
    let count = 0;
    let sum = 0;
    let sum_minus_k = null;
    
    for (let num of nums) {
        sum += num;
        sum_minus_k = sum - k;
        if (hash.hasOwnProperty(sum_minus_k)) {
            count += hash[sum_minus_k];
        }
        hash[sum] = hash.hasOwnProperty(sum) ? hash[sum] + 1 : 1;
    }
    return count;
}


var subarraySum2 = function(nums, k) {
    let count = 0;
    for (let i=0; i<nums.length; i++) {
        let currSum = 0;
        for (let j=i; j<nums.length; j++) {
            currSum += nums[j];
            if (currSum === k) {
                count++;
            }
        }
    }
    return count;
};